pkg/provisioner/dependency.go

```diff
@@ -191,3 +86,83 @@ func kind(tpl *bytes.Buffer, env v1alpha1.Environment) error {
 	}
 	return kind.Execute(tpl, env)
 }
+
+// DependencySolver is a struct that holds the dependency graph
+type DependencySolver struct {
+	Dependencies []ProvisionFunc
+	env          v1alpha1.Environment
+}
+
+func NewDependencySolver(env v1alpha1.Environment) *DependencySolver {
+	return &DependencySolver{
+		env: env,
+	}
+}
+
+func (d *DependencySolver) withKubernetes() {
+	switch d.env.Spec.Kubernetes.KubernetesInstaller {
+	case "kubeadm":
+		d.Dependencies = append(d.Dependencies, functions["kubeadm"])
+	case "kind":
+		d.Dependencies = append(d.Dependencies, functions["kind"])
+	case "microk8s":
+		// reset the graph to only include microk8s
+		d.Dependencies = nil
+		d.Dependencies = append(d.Dependencies, functions["microk8s"])
+	default:
+		// default to kubeadm if KubernetesInstaller is empty
+		d.Dependencies = append(d.Dependencies, functions["kubeadm"])
+	}
+}
+
+func (d *DependencySolver) withContainerRuntime() {
+	switch d.env.Spec.ContainerRuntime.Name {
+	case "containerd":
+		d.Dependencies = append(d.Dependencies, functions["containerd"])
+	case "crio":
+		d.Dependencies = append(d.Dependencies, functions["crio"])
+	case "docker":
+		d.Dependencies = append(d.Dependencies, functions["docker"])
+	default:
+		// default to containerd if ContainerRuntime.Name is empty
+		d.Dependencies = append(d.Dependencies, functions["containerd"])
+	}
+}
+
+func (d *DependencySolver) withContainerToolkit() {
+	d.Dependencies = append(d.Dependencies, functions["containerToolkit"])
+}
+
+func (d *DependencySolver) withNVDriver() {
+	d.Dependencies = append(d.Dependencies, functions["nvdriver"])
+}
+
+// buildDependencyGraph builds a dependency graph based on the environment
+// and returns a topologically sorted list of provisioning functions
+// to be executed in an opinionated order. We go bottom up in the graph
+// NVDriver -> Container Runtime -> Container Toolkit -> Kubernetes
+// Kubernetes -> Container Runtime -> Container Toolkit -> NVDriver
+// if a dependency is needed and not defined, we set an opinionated default
```

Comment:
What does `bottom-up` mean in this context? Looking at the example it seems as if ONLY NVDriver and Kubernetes are swapped. Is this correct?